lib/vnet/dns/dns.go

```diff
@@ -0,0 +1,391 @@
+		slog.DebugContext(ctx, "Matched DNS A.", "a", tcpip.AddrFrom4(result.A))
+		response, err = buildAResponse(buf, &requestHeader, &question, result.A)
+	case question.Type == dnsmessage.TypeAAAA && result.AAAA != ([16]byte{}):
+		slog.DebugContext(ctx, "Matched DNS AAAA.", "aaaa", tcpip.AddrFrom16(result.AAAA))
+		response, err = buildAAAAResponse(buf, &requestHeader, &question, result.AAAA)
+	default:
+		slog.DebugContext(ctx, "Forwarding unmatched query.")
+		return trace.Wrap(s.forward(ctx, slog, buf, responseWriter), "forwarding unmatched DNS query")
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = responseWriter.Write(response)
+	return trace.Wrap(err, "writing DNS response")
+}
+
+// forward forwards raw DNS messages to all upstream nameservers and writes the first response to
+// [responseWriter]. If there are no upstream nameservers, or none of them responds within the timeout, an
+// error is returned.
+func (s *Server) forward(ctx context.Context, slog *slog.Logger, buf []byte, responseWriter io.Writer) error {
+	ctx, cancel := context.WithTimeout(ctx, forwardRequestTimeout)
+	defer cancel()
+	deadline, _ := ctx.Deadline()
+
+	upstreamNameservers, err := s.upstreamNameserverSource.UpstreamNameservers(ctx)
+	if err != nil {
+		return trace.Wrap(err, "getting host default nameservers")
+	}
+	if len(upstreamNameservers) == 0 {
+		return trace.Errorf("no upstream nameservers")
+	}
+
+	// Forward the message to each upstream nameserver concurrently, the first to answer wins.
+	// Each goroutine will write a single error or a single response to the appropriate channel.
+	// Each goroutine should quickly exit after the context is canceled.
+	responses := make(chan []byte, len(upstreamNameservers))
+	errs := make(chan error, len(upstreamNameservers))
+	g, ctx := errgroup.WithContext(ctx)
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+	for _, nameserver := range upstreamNameservers {
+		responseBuf, returnResponseBuf := s.getMessageBuffer()
+		defer returnResponseBuf()
+
+		nameserver := nameserver
+		g.Go(func() error {
+			slog := slog.With("nameserver", nameserver)
+			slog.DebugContext(ctx, "Forwarding request to upstream nameserver.")
+
+			upstreamConn, err := net.Dial("udp", nameserver)
+			if err != nil {
+				errs <- trace.Wrap(err, "dialing upstream nameserver")
+				return nil
+			}
+
+			// Immediately close the upstream conn after the context is canceled to unblock any i/o. This
+			// function will not return any answer until all errgroup goroutines have terminated.
+			go func() {
+				<-ctx.Done()
+				upstreamConn.Close()
+			}()
+
+			upstreamConn.SetDeadline(deadline)
+			_, err = upstreamConn.Write(buf)
+			if err != nil {
+				errs <- trace.Wrap(err, "writing message to upstream")
+				return nil
+			}
+			n, err := upstreamConn.Read(responseBuf)
+			if err != nil {
+				errs <- trace.Wrap(err, "reading forwarded DNS response")
+				return nil
+			}
+			if n == cap(responseBuf) {
```

Comment:
why use `cap` instead of `len` here?  Seems brittle to assume that no future changes will ever result in use of a buffer whose capacity differs from its length.

Reply:
good point i'll switch this to `len`